source/common/upstream/cds_incremental.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <functional>
+
+#include "envoy/api/v2/cds.pb.h"
+#include "envoy/config/incremental_subscription.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/local_info/local_info.h"
+#include "envoy/stats/scope.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/common/logger.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * Incremental CDS API implementation that fetches via IncrementalSubscription.
+ */
+class CdsIncremental : public CdsApi,
+                       Config::IncrementalSubscriptionCallbacks<envoy::api::v2::Cluster>,
+                       Logger::Loggable<Logger::Id::upstream> {
+public:
+  static CdsApiPtr create(const envoy::api::v2::core::ConfigSource& cds_config,
```

Comment:
I don't think we need a completely new implementation here, we can probably share most of the code with the existing CdsApiImpl; what you would do is have `CdsApiImpl` as a subclass of `CdsApiIncrementalImpl`, it would translate the state-of-the-world `onConfigUpdate` into a parent call of `onIncrementalConfigUpdate`.

Reply:
Done. It got a little hairy. The code will be a little awkward until we have reimplemented general gRPC xDS in terms of incremental in this way. I'm wondering if it might be better to try to implement incremental ADS+general xDS all at once, if we are going to go the original-as-a-subclass-translation-wrapper route.

Reply:
It looks like I misunderstood you; I thought you meant to translate the non-incremental gRPC messages into incremental. Looking at just translating the `onConfigUpdate` callbacks to incremental, I don't see what there is to be shared - it looks like <10 lines of code, scattered across a few functions.

Reply:
Consolidated into one class.